okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,291 @@
+        if (firstUhe == null) {
+          firstUhe = uhe;
+        } else {
+          addSuppressedIfPossible(firstUhe, uhe);
+        }
+      }
+    }
+
+    if (firstUhe == null) {
+      throw new UnknownHostException("No lookup sources for host " + hostname);
+    }
+
+    throw firstUhe;
+  }
+
+  private List<InetAddress> lookupHttps(String hostname) throws UnknownHostException {
+    try {
+      ByteString query = DnsRecordCodec.encodeQuery(hostname, includeIPv6);
+
+      Request request = buildRequest(query);
+      Response response = executeRequest(request);
+
+      return readResponse(hostname, response);
+    } catch (UnknownHostException uhe) {
+      throw uhe;
+    } catch (Exception e) {
+      UnknownHostException unknownHostException = new UnknownHostException(hostname);
+      unknownHostException.initCause(e);
+      throw unknownHostException;
+    }
+  }
+
+  private Response executeRequest(Request request) throws IOException {
+    // cached request
+    if (!post && client.cache() != null) {
+      CacheControl cacheControl =
+          new CacheControl.Builder().maxStale(Integer.MAX_VALUE, TimeUnit.SECONDS).build();
+      Request cacheRequest = request.newBuilder().cacheControl(cacheControl).build();
+
+      Response response = client.newCall(cacheRequest).execute();
+
+      if (response.isSuccessful()) {
+        return response;
+      }
+    }
+
+    return client.newCall(request).execute();
+  }
+
+  private List<InetAddress> readResponse(String hostname, Response response) throws Exception {
+    if (response.cacheResponse() == null && response.protocol() != Protocol.HTTP_2) {
+      Platform.get().log(Platform.WARN, "Incorrect protocol: " + response.protocol(), null);
+    }
+
+    try {
+      if (!response.isSuccessful()) {
+        throw new IOException("response: " + response.code() + " " + response.message());
+      }
+
+      ResponseBody body = response.body();
+
+      if (body.contentLength() > MAX_RESPONSE_SIZE) {
+        throw new IOException("response size exceeds limit ("
+            + MAX_RESPONSE_SIZE
+            + " bytes): "
+            + body.contentLength()
+            + " bytes");
+      }
+
+      ByteString responseBytes = body.source().readByteString();
+
+      return DnsRecordCodec.decodeAnswers(hostname, responseBytes);
+    } finally {
+      response.close();
+    }
+  }
+
+  private Request buildRequest(ByteString query) {
+    Request.Builder requestBuilder = new Request.Builder().header("Accept", contentType.toString());
+
+    if (post) {
+      requestBuilder = requestBuilder.url(url).post(RequestBody.create(contentType, query));
+    } else {
+      String encoded = query.base64Url().replace("=", "");
+      HttpUrl requestUrl = url.newBuilder().addQueryParameter("dns", encoded).build();
+
+      requestBuilder = requestBuilder.url(requestUrl);
+    }
+
+    return requestBuilder.build();
+  }
+
+  static boolean isPrivateHost(String host) {
```

Comment:
I wonder if the simplest strategy for these would be for the builder to just have two more options: “resolvePrivateHosts(boolean)” and “resolvePublicHosts(boolean)”, defaulting to public-only